### PR TITLE
Implementation of #3

### PR DESCRIPTION
**Pull request by Lioncat2002**
#### Problem
- Implementation of issue #3 

#### Summary of Changes
- Added ConfirmTransaction using the rpc client
- Generating timing information between transaction send and confirmed

### CODE REVIEW

**Comment by Lioncat2002** [comment]
For the metrics
I need to save them into a file.
So should I save them in a csv file?
or a normal txt file?

File: light-rpc/metrics.csv [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+34.69397,31.256388,3.437581999999999
+34.355047,37.348667,2.99362
+35.543551,37.566298,2.0227470000000025
+33.739997,36.60322,2.863222999999998
+34.2413,35.888405,1.6471049999999963
+39.02204,39.16363,0.14159000000000077
+38.457204,36.360513,2.096691
+40.040714,34.705837,5.334876999999999
+38.79393,37.941486,0.8524440000000055
+42.978491,37.063746,5.914744999999996
```
Comment (mschneider) [inline light-rpc/metrics.csv:102]: please always include a header column.

we had a look at these results and they don't match our experience, it looks like everything got averaged

File: light-rpc/benches/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        .write(true)
+        .append(true)
+        .open("metrics.csv")
+        .unwrap();
+    let mut wtr = csv::Writer::from_writer(file);
+    b.iter(|| {
+        sig = forward_transaction_sender(&light_rpc, &alice, &bob, lamports);
+    });
+    let forwarder_time = (b.ns_per_iter() as f64) / 1000000 as f64;
+    b.iter(|| {
```
Comment (mschneider) [inline light-rpc/benches/metrics.rs:96]: very likely using the bencher class introduces some averaging, keep it simple

File: light-rpc/benches/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        confirm_transaction_sender(&light_rpc, &sig, 240);
+    });
+    let confirm_time = (b.ns_per_iter() as f64) / 1000000 as f64;
+
+    wtr.write_record(&[
+        forwarder_time.to_string(),
+        confirm_time.to_string(),
+        (forwarder_time - confirm_time).abs().to_string(),
+    ])
```
Comment (mschneider) [inline light-rpc/benches/metrics.rs:105]: don't write to files during a benchmark, your code should be structured manually into 3 phases:
1. setup - pre-allocate everything you will need, launch all threads you might need
2. a tight loop you are trying to benchmark - just do whatever you are trying to measure, try to work with minimal allocations to not end up influencing your measurement results
3. results - collect and analyze all results, write csv files for easy archival

Comment (mschneider) [reply light-rpc/benches/metrics.rs:105]: https://github.com/blockworks-foundation/solana/blob/max/bench-mango/bench-mango/src/main.rs#L636

Comment (godmodegalactus) [reply light-rpc/benches/metrics.rs:105]: I dont understand the significance of forwarder_time - confirm_time, should be more like forwarder_time + confirm_time.

**Review by godmodegalactus — CHANGES_REQUESTED** [review]

File: light-rpc/benches/metrics.rs (hunk 3 above)
Comment (Lioncat2002) [reply light-rpc/benches/metrics.rs]: > I dont understand the significance of forwarder_time - confirm_time, should be more like forwarder_time + confirm_time.


Umm I was trying to show like the duration between the time it took to confirm the transaction after the forward transaction sent back the signature

**Comment by aniketfuryrocks** [comment]
Please, merge #11 first. This merge will cause a lot of merge conflicts.

File: light-rpc/benches/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    let lamports = 1_000_000;
+    let mut data: Vec<Metrics> = vec![];
+
+    let mut wtr = csv::Writer::from_path("metrics.csv").unwrap();
+    let instant = SystemTime::now();
+    for _ in 0..times {
+        //generating a new keypair for each transaction
+        let start_time = instant.elapsed().unwrap().as_nanos();
+
+        let signatures = light_rpc::forward_transaction_sender(&light_rpc, lamports, 100);
```
Comment (godmodegalactus) [inline light-rpc/benches/metrics.rs:44]: Could you again seperate confirmation time and forwarding time.

**Review by godmodegalactus — APPROVED** [review]

File: light-rpc/benches/metrics.rs [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    std::{thread, time::Duration},
+};
+
+const RPC_ADDR: &str = "127.0.0.1:8899";
+const TPU_ADDR: &str = "127.0.0.1:1027";
+const CONNECTION_POOL_SIZE: usize = 1;
+
+#[derive(serde::Serialize)]
+struct Metrics {
+    #[serde(rename = "start time(ns)")]
```
Comment (godmodegalactus) [inline light-rpc/benches/metrics.rs]: Change it to ms instead

**Review by godmodegalactus — APPROVED** [review]

**Pull request merged**